typing: Move the typing notifications to below the mark as read banner.

roanster007 opened the pull request:
Previously, the typing notifications used to appear between the message list and the mark as read banner in a thread. This would cause the banner to shift down whenever someone starts typing, and shift back up if everyone stops typing.

This frequent bouncing of the banner could be distracting, and hence, is fixed by moving the typing notifications to below the mark as read banner.

<!-- Describe your pull request here.-->

[CZO thread](https://chat.zulip.org/#narrow/stream/9-issues/topic/Typing.20notifications.20cause.20bouncing.20banner/near/1778791)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/97049524/d3784ff5-bd20-41fb-b9fe-512a41e79f8b)<br>
![image](https://github.com/zulip/zulip/assets/97049524/1460a077-dac2-49d3-a8c9-e2636f8e29c6)



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
I think some margins need adjusting; this space is considerably bigger than what we had without the banner:

![image](https://github.com/zulip/zulip/assets/2746074/f391d633-7d82-4298-8f19-7ef5b810c2a7)

roanster007 commented:
@timabbott I adjusted the gap, and updated the screenshots.

timabbott commented on web/src/unread_ui.ts:
```diff
@@ -64,7 +65,8 @@ export function reset_unread_banner(): void {
     if (!user_closed_unread_banner) {
-        $("#mark_read_on_scroll_state_banner").toggleClass("invisible", false);
+        $("#mark_read_on_scroll_state_banner").toggleClass("hide", false);
+        $("#typing_notifications").css("margin-top", "-18px");
```
This should be done with new rules in the CSS files -- we basically never add CSS via `.css()`.

roanster007 replied:
oh! I git grepped .css() and found some files using that, so I thought I could use it. I'll update it.

roanster007 replied:
Updated!

roanster007 commented on web/styles/typing_notifications.css:
```diff
@@ -9,3 +9,7 @@
+
+.align_below_mark_as_read_banner {
+    margin-top: -18px;
+}
```
Not sure if this is the best place to have this rule here

sayamsamal replied:
Using a negative margin for this feels a bit off for me. We could use that if we don't have any other solutions, sure. But could you investigate why we have that margin below the "Mark as read" banner?

sayamsamal replied:
Okay, while this margin is important for the compose box, I don't see its purpose in the main message area, we should override it there. That should remove the need for negative margins.

roanster007 replied:
updated!

timabbott commented:
@sayamsamal can you help review this one?

roanster007 commented:
@sayamsamal gentle bump on this.

sayamsamal commented on web/src/unread_ui.ts:
```diff
@@ -54,7 +54,8 @@ export function update_unread_banner(): void {
 export function hide_unread_banner(): void {
     // Use visibility instead of hide() to prevent messages on the screen from
```
I think you missed this comment. We would want to remove this if we have shifted from "invisible" to "hide".

timabbott replied:
I'm not sure I quite understand the shift you're talking about, but we don't want weird unexpected.

sayamsamal replied:
I think @roanster007  is referring to the layout shift caused by the 'Mark as read' banner appearing between the 'Someone is typing...' text and the last message. I feel that is fine since 'Mark as unread' is a user-enabled action, a conscious choice and not something someone would do often.

roanster007 replied:
ya that's what I was referring to.

sayamsamal commented:
Sorry for the delay @roanster007, I have left you some feedback.

roanster007 commented:
Addressed reviews:

>Okay, while this margin is important for the compose box, I don't see its purpose in the main message area, we should override it there. That should remove the need for negative margins.

by overriding margin-bottom for `#mark_as_read_turned_off_banner` using `!important`

>I think you missed this comment. We would want to remove this if we have shifted from "invisible" to "hide".

by removing the comment.

roanster007 commented:
@sayamsamal this is ready for another review.

sayamsamal commented on web/src/unread_ui.ts:
```diff
@@ -64,7 +63,8 @@ export function reset_unread_banner(): void {
     if (!user_closed_unread_banner) {
-        $("#mark_read_on_scroll_state_banner").toggleClass("invisible", false);
+        $("#mark_read_on_scroll_state_banner").toggleClass("hide", false);
+        $("#typing_notifications").toggleClass("align_below_mark_as_read_banner", true);
```
This line of code is redundant?

roanster007 replied:
oh, thanks for catching that. updated!

sayamsamal commented on web/styles/typing_notifications.css:
```diff
@@ -9,3 +9,9 @@
+/*  override margine-bottom of main-view-banner since it causes gap
+    between mark as read banner and typing notification.  */
+#mark_as_read_turned_off_banner {
+    margin-bottom: 0 !important;
```
You don't need to use `!important` over here, the specificity of ID > Classes. Also, let's do `margin: 5px 0` instead, the typing notification feels too close to the banner.

roanster007 replied:
done!

sayamsamal replied:
The comment should no longer say `margin-bottom` and let's bring this comment just above the margin property since we are talking about that property specifically.

sayamsamal commented on web/src/unread_ui.ts:
```diff
@@ -52,9 +52,8 @@ export function update_unread_banner(): void {
-    // shifting vertically.
-    $("#mark_read_on_scroll_state_banner").toggleClass("invisible", true);
+    $("#mark_read_on_scroll_state_banner").toggleClass("hide", true);
+    $("#typing_notifications").toggleClass("align_below_mark_as_read_banner", false);
```
This line of code is also redundant.

roanster007 replied:
updated!

sayamsamal commented:
@roanster007 I have suggested some more changes.

roanster007 commented:
Updates:
>You don't need to use !important over here, the specificity of ID > Classes. Also, let's do margin: 5px 0 instead, the typing notification feels too close to the banner.

updated:
![image](https://github.com/zulip/zulip/assets/97049524/5a22b876-5cea-4a03-9cb2-b1f1e1b1dab2)
![image](https://github.com/zulip/zulip/assets/97049524/583a1e03-36a1-4861-b7f6-1ed193c04de7)

- Updated some redundant terms I overlooked.

sayamsamal commented:
Except for this minor mistype https://github.com/zulip/zulip/pull/29705#discussion_r1607335869, LGTM!

roanster007 commented:
@timabbott this is ready for your review. (addressed all reviews from sayam)

timabbott commented on web/src/unread_ui.ts:
```diff
@@ -52,9 +52,7 @@ export function update_unread_banner(): void {
-    // Use visibility instead of hide() to prevent messages on the screen from
-    // shifting vertically.
-    $("#mark_read_on_scroll_state_banner").toggleClass("invisible", true);
+    $("#mark_read_on_scroll_state_banner").toggleClass("hide", true);
```
This comment suggests the fact that this pushes down `bottom_whitespace` in some way; how does that work?

roanster007 replied:
so `invisible` just makes the element not visible, but it would still be present in the DOM. However, in case of `hide`, it would be removed from DOM as well.

In this case, when you use `hide` and scroll to the bottom of list, and mark message as unread, the `mark_as_read_banner` would cause the scroll position to slightly shift since a new element is coming up into document. However, in case of `invisible`, nothing would happen since element is already in document -- just that it is made visible.

timabbott replied:
I don't think that's the right terminology -- hide/visible have different behavior for whether it takes up space in the page layout, but the elements are present in the DOM in either case.

roanster007 replied:
seems like the `hide` is replaced with `invisible` first in 1b24bfcc3130efd22f8a38fca84a6f5364b55fd5, and the commit message specifies --
>mark as read: Fix banner not closing when x is clicked.
Previously, the X in the mark-as-read banner did nothing.

I don't seem to reproduce this with hide().

roanster007 replied:
According to [stackoverflow](https://stackoverflow.com/questions/133051/what-is-the-difference-between-visibilityhidden-and-displaynone):
> display:none means that the tag in question will not appear on the page at all (although you can still interact with it through the dom). There will be no space allocated for it between the other tags.
visibility:hidden means that unlike display:none, the tag is not visible, but space is allocated for it on the page. The tag is rendered, it just isn't seen on the page.

So the comment in the code - 
```
// Use visibility instead of hide() to prevent messages on the screen from
// shifting vertically.
````
maybe refers to shifting of message list due to allocation/deallocation of the space to the `mark as read` banner (in case of `visibility:hidden` since tag is rendered even when hidden, maybe we won't observer the shitft).

However, this issue was not part of #21595 where this was first introduced.

amanagr replied:
Since hiding (`display: none`) the banner instead of making it invisible will reduce scrollable space, it will cause the message list to shift dowards if the scrollbar is at bottom.

To avoid this we can add a `mark_read_on_scroll_state_banner_holder` div below typing notification div which just has the same dimesnsions as the unread banner before it is hidden. Just make sure to keep it hidden by default and hide it when user switches narrows or another unread banner is displayed.

amanagr replied:
Thanks for making that change. Works fine but we can possible make this look better by extracting the call:
```js
    $("#mark_read_on_scroll_state_banner_place_holder").toggleClass("hide", false);
    $("#mark_read_on_scroll_state_banner_place_holder").toggleClass("invisible", true);
```

and call it exactly what it does - `togge_dummy_banner` where we can pass the boolean set `true/false` for these classes.

roanster007 replied:
updated!

roanster007 commented:
> To avoid this we can add a mark_read_on_scroll_state_banner_place_holder div below typing notification div which just has the same dimesnsions as the unread banner before it is hidden. Just make sure to keep it hidden by default and hide it when user switches narrows or another unread banner is displayed.

@timabbott , as per Aman's suggestion, I added the placeholder.

amanagr commented on web/src/unread_ui.ts:
```diff
@@ -30,6 +30,13 @@ export function update_unread_banner(): void {
+    // This place holder is essential since hiding the unread banner would reduce
+    // scrollable place, shifting the message list downward if the scrollbar is
+    // at bottom.
+    $("#mark_read_on_scroll_state_banner_place_holder").html(render_mark_as_read_disabled_banner());
```
I like the idea of directly using `render_mark_as_read_disabled_banner()` here but we need to do it for other cases too since the text is different in them and hence the height.

So, you can extract a function which sets the html for both the place holder and visible banner.

roanster007 replied:
I updated to use the existing if-else conditions used to determine innerhtml for `#mark_read_on_scroll_state_banner`.

amanagr commented on web/styles/typing_notifications.css:
```diff
@@ -9,3 +9,9 @@
 }
+
+#mark_as_read_turned_off_banner {
+    /*  override the margin of main-view-banner since it causes gap
```
`causes gap` -> `causes more gap than we want`

Also, looks like we need this for other banners as well which are displayed in the message feed.

roanster007 replied:
> causes gap -> causes more gap than we want

updated
> Also, looks like we need this for other banners as well which are displayed in the message feed.
there are only some indicators (no banners)

 other than the `mark_read_on_scroll_state_banner` there's only `scheduled_message_indicator` below `message-lists-container`, which deosn't seem to impact it anyways.

amanagr replied:
I see, thanks!

amanagr commented:
@roanster007 thanks for the update! Left some comments!

roanster007 commented:
@amanagr could you have another look at it.

amanagr commented on web/src/unread_ui.ts:
```diff
@@ -30,11 +30,16 @@ export function update_unread_banner(): void {
         user_settings.web_mark_read_on_scroll_policy ===
         web_mark_read_on_scroll_policy_values.never.code
     ) {
         $("#mark_read_on_scroll_state_banner").html(render_mark_as_read_disabled_banner());
+        $("#mark_read_on_scroll_state_banner_place_holder").html(
+            render_mark_as_read_disabled_banner(),
+        );
```
Let's extract all these `html` calls into a function which just takes the rendered template as input.

roanster007 replied:
Added a prep commit, which extracts them into a separate method.

amanagr commented:
@roanster007 left a minor comment.  Looks good to me otherwise.

alya commented:
@timabbott Let me know if I need to test.

timabbott commented on web/styles/typing_notifications.css:
```diff
@@ -9,3 +9,10 @@
+#mark_as_read_turned_off_banner {
+    /*  override the margin of main-view-banner since it causes
+    more gap than we want between mark as read banner and typing
+    notification.  */
```
This is not our standard whitespace pattern for comments in CSS.

roanster007 replied:
oops, updated

timabbott commented on web/src/unread_ui.ts:
```diff
@@ -23,8 +23,12 @@ export function register_update_unread_counts_hook(f: UpdateUnreadCountsHook): v
+    // This place holder is essential since hiding the unread banner would reduce
+    // scrollable place, shifting the message list downward if the scrollbar is
+    // at bottom.
+    $("#mark_read_on_scroll_state_banner_place_holder").html(rendered_banner);
```
Why does the same HTML go into both places? I don't really understand.

roanster007 replied:
>To avoid this we can add a mark_read_on_scroll_state_banner_place_holder div below typing notification div which just has the same dimesnsions as the unread banner before it is hidden. 

According to Aman's comment, we want same dimensions both the actual banner and the placeholder banner. So both of them are being rendered with the same template so that the dimensions are adjusted automatically.

roanster007 commented on web/src/unread_ui.ts:
```diff
@@ -23,6 +23,10 @@ export function register_update_unread_counts_hook(f: UpdateUnreadCountsHook): v
 }
 
+function set_mark_read_on_scroll_state_banner(template: string): void {
+    $("#mark_read_on_scroll_state_banner").html(template);
```
updated to `banner_html` from `rendered_banner`

timabbott commented:
OK, this seems like it should work, merged, thanks @roanster007 and @amanagr!
